README.md
# haskell-tick-tack-roll

Tick tack roll game in haskell

## Getting started

To make it easy for you to get started with GitLab, here's a list of recommended next steps.

Already a pro? Just edit this README.md and make it your own. Want to make it easy? [Use the template at the bottom](#editing-this-readme)!

## Add your files

- [ ] [Create](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#create-a-file) or [upload](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#upload-a-file) files
- [ ] [Add files using the command line](https://docs.gitlab.com/ee/gitlab-basics/add-file.html#add-a-file-using-the-command-line) or push an existing Git repository with the following command:

```
cd existing_repo
git remote add origin https://git.gvk.idi.ntnu.no/course/prog2006/as/jonask/haskell-tick-tack-roll.git
git branch -M main
git push -uf origin main
```

## Integrate with your tools

- [ ] [Set up project integrations](https://git.gvk.idi.ntnu.no/course/prog2006/as/jonask/haskell-tick-tack-roll/-/settings/integrations)

## Collaborate with your team

- [ ] [Invite team members and collaborators](https://docs.gitlab.com/ee/user/project/members/)
- [ ] [Create a new merge request](https://docs.gitlab.com/ee/user/project/merge_requests/creating_merge_requests.html)
- [ ] [Automatically close issues from merge requests](https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
- [ ] [Enable merge request approvals](https://docs.gitlab.com/ee/user/project/merge_requests/approvals/)
- [ ] [Automatically merge when pipeline succeeds](https://docs.gitlab.com/ee/user/project/merge_requests/merge_when_pipeline_succeeds.html)

## Test and Deploy

Use the built-in continuous integration in GitLab.

- [ ] [Get started with GitLab CI/CD](https://docs.gitlab.com/ee/ci/quick_start/index.html)
- [ ] [Analyze your code for known vulnerabilities with Static Application Security Testing(SAST)](https://docs.gitlab.com/ee/user/application_security/sast/)
- [ ] [Deploy to Kubernetes, Amazon EC2, or Amazon ECS using Auto Deploy](https://docs.gitlab.com/ee/topics/autodevops/requirements.html)
- [ ] [Use pull-based deployments for improved Kubernetes management](https://docs.gitlab.com/ee/user/clusters/agent/)
- [ ] [Set up protected environments](https://docs.gitlab.com/ee/ci/environments/protected_environments.html)

***

# Editing this README

When you're ready to make this README your own, just edit this file and use the handy template below (or feel free to structure it however you want - this is just a starting point!).  Thank you to [makeareadme.com](https://www.makeareadme.com/) for this template.

## Suggestions for a good README
Every project is different, so consider which of these sections apply to yours. The sections used in the template are suggestions for most open source projects. Also keep in mind that while a README can be too long and detailed, too long is better than too short. If you think your README is too long, consider utilizing another form of documentation rather than cutting out information.

## Name
Choose a self-explaining name for your project.

## Description
Let people know what your project can do specifically. Provide context and add a link to any reference visitors might be unfamiliar with. A list of Features or a Background subsection can also be added here. If there are alternatives to your project, this is a good place to list differentiating factors.

## Badges
On some READMEs, you may see small images that convey metadata, such as whether or not all the tests are passing for the project. You can use Shields to add some to your README. Many services also have instructions for adding a badge.

## Visuals
Depending on what you are making, it can be a good idea to include screenshots or even a video (you'll frequently see GIFs rather than actual videos). Tools like ttygif can help, but check out Asciinema for a more sophisticated method.

## Installation
Within a particular ecosystem, there may be a common way of installing things, such as using Yarn, NuGet, or Homebrew. However, consider the possibility that whoever is reading your README is a novice and would like more guidance. Listing specific steps helps remove ambiguity and gets people to using your project as quickly as possible. If it only runs in a specific context like a particular programming language version or operating system or has dependencies that have to be installed manually, also add a Requirements subsection.

## Usage
Use examples liberally, and show the expected output if you can. It's helpful to have inline the smallest example of usage that you can demonstrate, while providing links to more sophisticated examples if they are too long to reasonably include in the README.

## Support
Tell people where they can go to for help. It can be any combination of an issue tracker, a chat room, an email address, etc.

## Roadmap
If you have ideas for releases in the future, it is a good idea to list them in the README.

## Contributing
State if you are open to contributions and what your requirements are for accepting them.

For people who want to make changes to your project, it's helpful to have some documentation on how to get started. Perhaps there is a script that they should run or some environment variables that they need to set. Make these steps explicit. These instructions could also be useful to your future self.

You can also document commands to lint the code or run tests. These steps help to ensure high code quality and reduce the likelihood that the changes inadvertently break something. Having instructions for running tests is especially helpful if it requires external setup, such as starting a Selenium server for testing in a browser.

## Authors and acknowledgment
Show your appreciation to those who have contributed to the project.

## License
For open source projects, say how it is licensed.

## Project status
If you have run out of energy or time for your project, put a note at the top of the README saying that development has slowed down or stopped completely. Someone may choose to fork your project or volunteer to step in as a maintainer or owner, allowing your project to keep going. You can also make an explicit request for maintainers.


ChangeLog.md
# Changelog for haskell-tick-tack-roll

## Unreleased changes


test/Spec.hs
-- Testing dependencies
import Test.DocTest(doctest) 
import Test.QuickCheck(quickCheck) 
import Test.Hspec(Spec, hspec, describe, shouldBe, it) 
import Test.Hspec.QuickCheck(prop) 

-- Functions to test
import Roll -- Rolling functions
import Lib -- Other functions

main :: IO ()
main = do
    putStrLn "Doctests:"
    doctest ["-isrc", "app/Main.hs"]
    putStrLn "\nHspec tests:"
    -- Tests are arranged by dependendy order, so later tests functions depend on functions from earlier ones.
    hspec $ do
        testSquaredInteger
        testLengthCheck
        testSwap
        testRollRowLeft
        testRollRowsHelper
        testRollHelper
        testRoll
        

testSquaredInteger :: Spec
testSquaredInteger =
    describe "\nTests for squaredInteger function" $ do
        it "nine is 3 to the power of two" $ do
           squaredInteger 9 `shouldBe` 3
        it "16 is four to the power of two" $ do
            squaredInteger 16 `shouldBe` 4
        it "25 is five to the power of two" $ do
            squaredInteger 25 `shouldBe` 5
        it "Rounds the squared number if it doesn't return an integer" $ do
            squaredInteger 14 `shouldBe` 4
        it "Handles zero value gracefully" $ do
            squaredInteger 0 `shouldBe` 0
        it "Handles negatives gracefully" $ do
            squaredInteger (-25) `shouldBe` 0


testLengthCheck :: Spec
testLengthCheck =
    describe "\nTests for lengthCheck function" $ do
        it "lengthCheck valid 3x3 grid" $ do
           lengthCheck [1..9] `shouldBe` 3
        it "lengthCheck valid 5x5 grid" $ do
            lengthCheck [1..25] `shouldBe` 5
        it "lengthCheck invalid (too small) grid" $ do
            lengthCheck [1..4] `shouldBe` 0
        it "lengthCheck invalid (not odd number) grid" $ do
            lengthCheck [1..16] `shouldBe` 0
        it "lengthCheck invalid (not squareable into an integer) grid" $ do
            lengthCheck [1..37] `shouldBe` 0


testSwap :: Spec
testSwap =
    describe "\nTests for swap function" $ do
        it "3x3 grid should swap item one and three" $ do
            swap [1..9] `shouldBe` [3,2,1,4,5,6,7,8,9]
        it "5x5 grid should swap item one and five" $ do
            swap [1..25] `shouldBe` [5,2,3,4,1,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
        it "4x4 grid is invalid and should return an empty array" $ do
            swap [1..16] `shouldBe` []
        it "empty input should produce empty output" $ do
            swap ([]::[Int]) `shouldBe` []


testRollRowLeft :: Spec
testRollRowLeft =
    describe "\nTests for rollRowLeft function" $ do
        it "rollRowLeft 3x3 grid first row" $ do
           rollRowLeft 3 0 0 [1..9] `shouldBe` [1,4,7]
        it "rollRowLeft left 5x5 grid first row" $ do
            rollRowLeft 5 0 0 [1..25] `shouldBe` [1,6,11,16,21]


-- | testRollRowsHelper test for rollRowsHelper function
testRollRowsHelper :: Spec
testRollRowsHelper =
    describe "\nTests for rollRowsHelper function" $ do
        it "rollRowsHelper left 3x3 grid" $ do
           rollRowsHelper [1..9] True [0..2] 3 `shouldBe` [1,4,7,2,5,8,3,6,9]
        it "rollRowsHelper left 5x5 grid" $ do
            rollRowsHelper [1..25] True [0..4] 5 `shouldBe` [1,6,11,16,21,2,7,12,17,22,3,8,13,18,23,4,9,14,19,24,5,10,15,20,25]
        it "rollRowsHelper right 3x3 grid" $ do
            rollRowsHelper [1..9] False (reverse [0..2]) 3 `shouldBe` [7,4,1,8,5,2,9,6,3]
        it "rollRowsHelper right 5x5 grid" $ do
            rollRowsHelper [1..25] False (reverse [0..4]) 5 `shouldBe` [21,16,11,6,1,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,5]
        {- Add checking for empty grids or ranges? range should be of len rowLen, grid should be of len rowLen^2
        it "rollRowsHelper left no grid" $ do
            rollRowsHelper ([]::[Int]) False [0..2] 3 `shouldBe` []
        it "rollRowsHelper left no range 3x3 grid" $ do
            rollRowsHelper [1..9] False ([]::[Int]) 3 `shouldBe` []
        -}


testRollHelper :: Spec
testRollHelper =
    describe "\nTests for rollHelper function" $ do
        it "3x3 grid left" $ do
            rollHelper 3 True [1..9] `shouldBe` [3,6,9,2,5,8,1,4,7]
        it "5x5 grid left" $ do
            rollHelper 5 True [1..25] `shouldBe` [5,10,15,20,25,4,9,14,19,24,3,8,13,18,23,2,7,12,17,22,1,6,11,16,21]
        it "3x3 grid right" $ do
            rollHelper 3 False [1..9] `shouldBe` [7,4,1,8,5,2,9,6,3]
        it "5x5 grid right" $ do
            rollHelper 5 False [1..25] `shouldBe` [21,16,11,6,1,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,5]
        it "empty input array should produce empty output" $ do
            rollHelper 3 False ([]::[Int]) `shouldBe` []

-- | testRoll test for roll function
testRoll :: Spec
testRoll =
    describe "Tests for roll function" $ do
        it "Left roll 3x3 grid" $ do
            roll "left" [1..9] `shouldBe` [1,6,9,2,5,8,3,4,7]
        it "Left roll 5x5 grid" $ do
            roll "left" [1..25] `shouldBe` [1,10,15,20,25,4,9,14,19,24,3,8,13,18,23,2,7,12,17,22,5,6,11,16,21]
        it "Right roll 3x3 grid" $ do
            roll "right" [1..9] `shouldBe` [7,4,3,8,5,2,9,6,1]
        it "Right roll 5x5 grid" $ do
            roll "right" [1..25] `shouldBe` [21,16,11,6,5,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,1]
        it "Roll no direction 3x3 grid" $ do
            roll "" [1..9] `shouldBe` []
        it "Invalid grid size" $ do
            roll "right" [1..16] `shouldBe` []
        it "Roll right no grid" $ do
            roll "right" ([]::[Int]) `shouldBe` []

-- | testWinCheck test for winCheck function
testWinCheck :: Spec
testWinCheck =
    describe "Tests for WinCheck function" $ do
        it "3x3 board top-left->bottom-right" $ do
            winCheck ['X','O','_',
                      '_','X','O',
                      '_','_','X'] `shouldBe` 'X'
        it "3x3 board top-right->bottom-left" $ do
            winCheck ['O','X','O',
                      '_','O','X',
                      'O','_','X'] `shouldBe` 'O'
        it "3x3 board top-left->bottom-left" $ do
            winCheck ['X','O','O',
                      'X','O','X',
                      'X','_','X'] `shouldBe` 'X'
        it "3x3 board top-right->bottom-right" $ do
            winCheck ['X','X','O',
                      '_','X','O',
                      'X','_','O'] `shouldBe` 'O'
        it "3x3 board bottom-left->bottom-right" $ do
            winCheck ['X','O','O',
                      'O','O','X',
                      'X','X','X'] `shouldBe` 'X'
        it "3x3 board top-left->top-right" $ do
            winCheck ['O','O','O',
                      '_','X','X',
                      'X','_','X'] `shouldBe` 'O'
        it "3x3 board no win pattern" $ do
            winCheck ['O','X','O',
                      'O','X','X',
                      'X','O','X'] `shouldBe` 'O'
        it "invalid board" $ do
            winCheck ([]::[Char]) `shouldBe` '_'
        


src/Lib.hs
module Lib
    ( printBoard,
      winCheck
    ) where

import Roll -- For checking if the board size is valid

-- | printBoard takes a board and prints it to the screen, error message if the board is invalid
-- uses the helper function printLineByLine to print each line of the array
printBoard :: [Int] -> IO()
printBoard arr = do
    let len = lengthCheck arr
    if len /= 0 then do
        printLineByLine len arr
    else do
        putStrLn "Error: unable to print board due to invalid board size"

-- | printLineByLine takes a line length and a line list and prints the list line by line
-- uses the printLine helper function to print each line
printLineByLine :: Int -> [Int] -> IO()
printLineByLine len arr = do
    if length arr >= len then do
        putStr "# "
        printLine (take len arr)
        printLineByLine len (drop len arr)
    else
        return ()

-- | printLine takes an array of integers and prints them one by one separated by a single space
-- when done also print a line break
printLine :: [Int] -> IO()
printLine i = do
    if (length i) > 0 then do
        putStr (show (head i))
        putChar ' '
        printLine (tail i)
    else
        putStr "\n"

-- | winCheck takes a grid, determines the grid size and player pieces, returns the winner
-- or blank if no winner in the current board state
winCheck :: [Char] -> Char
winCheck a = do
    -- get the line length
    let len = lengthCheck a
    if len /= 0 then do
        -- make a list of the left diagonal and check the item count of X and O
        
        -- make a list of the right diagonal and check the item count of X and O

        -- check line by line
        let result = winCheckHorizontal len a -- Check horizontal lines
        if result /= '_' then
            result 
        else do -- Check vertical lines
            let result = winCheckHorizontal len (rollHelper len True a)
            result
    else -- Default to blank output if invalid input
        '_'

-- >>> winCheckHorizontal 3 ['X','X','X','O','X','O','X','_','X']
-- 'X'
--
-- >>> winCheckHorizontal 3 ['O','O','O','X','O','X','X','_','X']
-- 'O'
--

-- | winCheckHorizontal takes a grid size and a grid, returns the winner or '_', for horizontal wins
-- recurses until empty, if no winner before empty then return '_' 
winCheckHorizontal :: Int -> [Char] -> Char
winCheckHorizontal x a = do
    -- Take a line
    let line = take x a
    let winner = membersInListHelper x line
    if winner /= '_' || length a == x  then
        winner
    else
        winCheckHorizontal x (drop x a)

--winCheckDiagonal :: Bool -> [Char] -> Char


-- | membersInListhelper, helper function for membersInList, 
-- calls it with both characters and returns a victor if there is one
membersInListHelper :: Int -> [Char] -> Char
membersInListHelper x a = do
    if membersInList 'X' a == x then
        'X'
    else if membersInList 'O' a == x then
        'O'
    else
        '_'
    

-- | membersInList, uses list comprehension to go through each element of the input list, comparing it to
-- the signature and returning the number of occurrences
-- same as mySignature function from haskell-01
membersInList :: Eq a => a -> [a] -> Int
membersInList z xs = sum [if z == x then 1 else 0 | x <- xs]

src/Roll.hs
module Roll
    ( swap,
      roll,
      lengthCheck,
      rollHelper,
      rollRowsHelper,
      rollRowLeft,
      squaredInteger
    ) where

-- | swap tests, swaps the first and last element of the first line in valid grids
-- >>> swap [1..9]
-- [3,2,1,4,5,6,7,8,9]
--
-- >>> swap [1..25]
-- [5,2,3,4,1,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]
--
-- >>> swap [1..3]
-- []
--

-- | swap swaps the first and last mark off the first row
swap :: [a] -> [a]
swap x = do
    let len = lengthCheck x
    if len /= 0 then do
        -- swap item 0 and len in the list
        -- Get the first line, remove the head and tail (the items to swap)
        let middleFirstLine = init (tail (take len x))
        -- Get the rest of the list (drop the first line)
        let rest = drop len x
        -- Concattenate the lists
        [x!!(len-1)] ++ middleFirstLine ++ [x!!0] ++ rest
    else
        -- Return an empty array
        []   

-- | roll tests, takes a direction and a grid, rolls the grid the direction
-- >>> roll "left" [1..9]
-- [1,6,9,2,5,8,3,4,7]
--
-- >>> roll "left" [1..25]
-- [1,10,15,20,25,4,9,14,19,24,3,8,13,18,23,2,7,12,17,22,5,6,11,16,21]
--
-- >>> roll "right" [1..9]
-- [7,4,3,8,5,2,9,6,1]
--
-- >>> roll "right" [1..25]
-- [21,16,11,6,5,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,1]
--
-- >>> roll "" [1..25]
-- []
--
-- >>> roll "right" []
-- []
--
-- >>> roll "" []
-- []
--

-- roll function that rolls the grid the specified direction, returns an empty array on failure
roll :: String -> [a] -> [a]
roll s arr = do
    let len = lengthCheck arr
    if len /= 0 then do
        if s == "left" then
            rollHelper len True (swap arr)
        else if s == "right" then
            rollHelper len False (swap arr)
        else
            -- Return an empty array
            []
    else
        -- Return an empty array
        []   

-- | rollHelper tests, takes a direction and a grid, returns the grid rotated that direction
-- >>> rollHelper 3 True [1..9]
-- [3,6,9,2,5,8,1,4,7]
--
-- >>> rollHelper 5 False [1..25]
-- [21,16,11,6,1,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,5]
--
-- >>> rollHelper 0 False []
-- []
--

-- rollHelper helper function that rolls the grid the provided direction
rollHelper :: Int -> Bool -> [a] -> [a]
rollHelper _ _ [] = []
rollHelper len dir arr = do
    -- Rotate the array and return it
    if dir then do
        -- calculate the offset
        let offset = reverse [0..(len-1)]
        -- rotate the rows
        rollRowsHelper arr dir offset len
    else do
        -- calculate the offset
        let offset = reverse [0..(len-1)]
        -- rotate the rows
        rollRowsHelper arr dir offset len
    

-- | rollRowsHelper rolls all rows of an array left if true, right if false
-- >>> rollRowsHelper [1..9] True [0..2] 3
-- [1,4,7,2,5,8,3,6,9]
--
-- >>> rollRowsHelper [1..9] False (reverse [0..2]) 3
-- [7,4,1,8,5,2,9,6,3]
--
-- >>> rollRowsHelper [1..25] True [0..4] 5
-- [1,6,11,16,21,2,7,12,17,22,3,8,13,18,23,4,9,14,19,24,5,10,15,20,25]
--
-- >>> rollRowsHelper [1..25] False (reverse [0..4]) 5
-- [21,16,11,6,1,22,17,12,7,2,23,18,13,8,3,24,19,14,9,4,25,20,15,10,5]
--


-- | rollRowsHelper takes a grid and a offset list, returns the grid rotated, left if dir, else right
rollRowsHelper :: [a] -> Bool -> [Int] -> Int -> [a]
rollRowsHelper arr dir offset rowLen = do
    if length offset /= 0 then
        if dir then
            (rollRowLeft rowLen (head offset) 0 arr) ++ (rollRowsHelper arr dir (tail offset) rowLen)
        else
            (rollRowsHelper arr dir (tail offset) rowLen) ++ reverse (rollRowLeft rowLen (head  offset) 0 arr) 
    else
        []

-- | rollRowLeft tests, rolls a single roll left
-- >>> rollRowLeft 3 0 0 [1..9]
-- [1,4,7]
--
-- >>> rollRowLeft 3 1 0 [1..9]
-- [2,5,8]
--
-- >>> rollRowLeft 3 2 0 [1..9]
-- [3,6,9]
--


-- | rollRowLeft recursive function that rolls a single row left
-- takes the row length and the whole grid, a deviation and a recursion counter as variables
rollRowLeft :: Int -> Int -> Int -> [a] -> [a]
rollRowLeft rowLen offSet recCount arr = do
    if rowLen == recCount then
        []
    else
        [arr!!((rowLen*recCount)+offSet)] ++ rollRowLeft rowLen offSet (recCount+1) arr


-- | lengthCheck tests, works only with valid grid sizes (3x3 + 2x) where x is a whole number
-- if the length is less than 9, not squarable as a whole number, or not even 0 is returned.
-- >>> lengthCheck [1..9]
-- 3
--
-- >>> lengthCheck [1..25]
-- 5
--
-- >>> lengthCheck [1..11]
-- 0
--
-- >>> lengthCheck [1..16]
-- 0
--
-- >>> lengthCheck []
-- 0
--
-- >>> lengthCheck [1..3]
-- 0
--


-- | lengthCheck checks if the length is valid, returns the length if valid, 0 if not
-- if the length is less than 9, not squarable as a whole number, or not even 0 is returned.
-- else the length squared is returned
lengthCheck :: [a] -> Int
lengthCheck x = do
    let len = length x
    if len < 9 then
        0
    else if (len `mod` 2) /= 1 then
        0
    else do
        -- Check if the length squared is a whole number
        let rowLen = squaredInteger len
        -- If not a whole number return 0
        if (rowLen * rowLen) /= len then
            0
        -- Else return the row length
        else
            rowLen

-- | squared integer tests, works with positive, negative and zero values.
-- >>> squaredInteger 0
-- 0
--
-- >>> squaredInteger 2
-- 1
--
-- >>> squaredInteger 9
-- 3
--
-- >>> squaredInteger 16
-- 4
--
-- >>> squaredInteger (-10)
-- 0
--


-- | squaredInteger takes an integer and returns the integer squared rounded to an integer
squaredInteger :: Int -> Int
squaredInteger x = do
    if x > 0 then
        round . sqrt . fromIntegral $ x
    else
        0

-- inspired by the first comment to the solution from kqr at stackoverflow:
-- The comment: "How does it compare to intSqrt = floor . sqrt . fromInteger ? – kqr Nov 14, 2013 at 7:41"
-- Link: https://stackoverflow.com/questions/19965149/integer-square-root-function-in-haskell